<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!--NewPage-->
<HTML>
<HEAD>
<!-- Generated by javadoc (build 1.5.0_18) on Tue Jul 07 10:40:32 PDT 2009 -->
<TITLE>
Transcript
</TITLE>

<META NAME="keywords" CONTENT="apollo.datamodel.Transcript class">

<LINK REL ="stylesheet" TYPE="text/css" HREF="../../stylesheet.css" TITLE="Style">

<SCRIPT type="text/javascript">
function windowTitle()
{
    parent.document.title="Transcript";
}
</SCRIPT>
<NOSCRIPT>
</NOSCRIPT>

</HEAD>

<BODY BGCOLOR="white" onload="windowTitle();">


<!-- ========= START OF TOP NAVBAR ======= -->
<A NAME="navbar_top"><!-- --></A>
<A HREF="#skip-navbar_top" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_top_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Class</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;<A HREF="../../apollo/datamodel/SyntenyRegion.html" title="class in apollo.datamodel"><B>PREV CLASS</B></A>&nbsp;
&nbsp;<A HREF="../../apollo/datamodel/TranslationI.html" title="interface in apollo.datamodel"><B>NEXT CLASS</B></A></FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../index.html?apollo/datamodel/Transcript.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="Transcript.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
<TR>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
  SUMMARY:&nbsp;NESTED&nbsp;|&nbsp;<A HREF="#field_summary">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_summary">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_summary">METHOD</A></FONT></TD>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
DETAIL:&nbsp;<A HREF="#field_detail">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_detail">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_detail">METHOD</A></FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_top"></A>
<!-- ========= END OF TOP NAVBAR ========= -->

<HR>
<!-- ======== START OF CLASS DATA ======== -->
<H2>
<FONT SIZE="-1">
apollo.datamodel</FONT>
<BR>
Class Transcript</H2>
<PRE>
java.lang.Object
  <IMG SRC="../../resources/inherit.gif" ALT="extended by "><A HREF="../../apollo/datamodel/Range.html" title="class in apollo.datamodel">apollo.datamodel.Range</A>
      <IMG SRC="../../resources/inherit.gif" ALT="extended by "><A HREF="../../apollo/datamodel/SeqFeature.html" title="class in apollo.datamodel">apollo.datamodel.SeqFeature</A>
          <IMG SRC="../../resources/inherit.gif" ALT="extended by "><A HREF="../../apollo/datamodel/FeatureSet.html" title="class in apollo.datamodel">apollo.datamodel.FeatureSet</A>
              <IMG SRC="../../resources/inherit.gif" ALT="extended by "><A HREF="../../apollo/datamodel/AnnotatedFeature.html" title="class in apollo.datamodel">apollo.datamodel.AnnotatedFeature</A>
                  <IMG SRC="../../resources/inherit.gif" ALT="extended by "><B>apollo.datamodel.Transcript</B>
</PRE>
<DL>
<DT><B>All Implemented Interfaces:</B> <DD><A HREF="../../apollo/datamodel/AnnotatedFeatureI.html" title="interface in apollo.datamodel">AnnotatedFeatureI</A>, <A HREF="../../apollo/datamodel/Comparable.html" title="interface in apollo.datamodel">Comparable</A>, <A HREF="../../apollo/datamodel/FeatureSetI.html" title="interface in apollo.datamodel">FeatureSetI</A>, <A HREF="../../apollo/datamodel/RangeI.html" title="interface in apollo.datamodel">RangeI</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html" title="interface in apollo.datamodel">SeqFeatureI</A>, <A HREF="../../apollo/datamodel/TranslationI.html" title="interface in apollo.datamodel">TranslationI</A>, java.io.Serializable, java.lang.Cloneable</DD>
</DL>
<HR>
<DL>
<DT><PRE>public class <B>Transcript</B><DT>extends <A HREF="../../apollo/datamodel/AnnotatedFeature.html" title="class in apollo.datamodel">AnnotatedFeature</A></DL>
</PRE>

<P>
<DL>
<DT><B>See Also:</B><DD><A HREF="../../serialized-form.html#apollo.datamodel.Transcript">Serialized Form</A></DL>
<HR>

<P>
<!-- =========== FIELD SUMMARY =========== -->

<A NAME="field_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Field Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;<A HREF="../../apollo/datamodel/SequenceI.html" title="interface in apollo.datamodel">SequenceI</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../apollo/datamodel/Transcript.html#cDNA">cDNA</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected static&nbsp;org.apache.log4j.Logger</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../apollo/datamodel/Transcript.html#logger">logger</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;<A HREF="../../apollo/datamodel/SequenceI.html" title="interface in apollo.datamodel">SequenceI</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../apollo/datamodel/Transcript.html#peptide">peptide</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../apollo/datamodel/Transcript.html#TRANSCRIPT_TYPE">TRANSCRIPT_TYPE</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;<A NAME="fields_inherited_from_class_apollo.datamodel.AnnotatedFeature"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left"><B>Fields inherited from class apollo.datamodel.<A HREF="../../apollo/datamodel/AnnotatedFeature.html" title="class in apollo.datamodel">AnnotatedFeature</A></B></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><A HREF="../../apollo/datamodel/AnnotatedFeature.html#annotationRoot">annotationRoot</A>, <A HREF="../../apollo/datamodel/AnnotatedFeature.html#owner">owner</A></CODE></TD>
</TR>
</TABLE>
&nbsp;<A NAME="fields_inherited_from_class_apollo.datamodel.FeatureSet"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left"><B>Fields inherited from class apollo.datamodel.<A HREF="../../apollo/datamodel/FeatureSet.html" title="class in apollo.datamodel">FeatureSet</A></B></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><A HREF="../../apollo/datamodel/FeatureSet.html#features">features</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#flags">flags</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#genericReadThroughStopResidue">genericReadThroughStopResidue</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#hitSequence">hitSequence</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#minus1_frameshift">minus1_frameshift</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#missing_3prime">missing_3prime</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#missing_5prime">missing_5prime</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#plus1_frameshift">plus1_frameshift</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#POLYA_REMOVED">POLYA_REMOVED</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#readthrough_stop">readthrough_stop</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#standard_start_codon">standard_start_codon</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#start_codon">start_codon</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#trans_spliced">trans_spliced</A></CODE></TD>
</TR>
</TABLE>
&nbsp;<A NAME="fields_inherited_from_class_apollo.datamodel.SeqFeature"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left"><B>Fields inherited from class apollo.datamodel.<A HREF="../../apollo/datamodel/SeqFeature.html" title="class in apollo.datamodel">SeqFeature</A></B></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><A HREF="../../apollo/datamodel/SeqFeature.html#biotype">biotype</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#edit_offset_adjust">edit_offset_adjust</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#id">id</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#phase">phase</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#properties">properties</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#ref_features">ref_features</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#refFeature">refFeature</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#refId">refId</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#score">score</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#scores">scores</A></CODE></TD>
</TR>
</TABLE>
&nbsp;<A NAME="fields_inherited_from_class_apollo.datamodel.Range"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left"><B>Fields inherited from class apollo.datamodel.<A HREF="../../apollo/datamodel/Range.html" title="class in apollo.datamodel">Range</A></B></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><A HREF="../../apollo/datamodel/Range.html#high">high</A>, <A HREF="../../apollo/datamodel/Range.html#low">low</A>, <A HREF="../../apollo/datamodel/Range.html#name">name</A>, <A HREF="../../apollo/datamodel/Range.html#refSeq">refSeq</A>, <A HREF="../../apollo/datamodel/Range.html#strand">strand</A>, <A HREF="../../apollo/datamodel/Range.html#type">type</A></CODE></TD>
</TR>
</TABLE>
&nbsp;<A NAME="fields_inherited_from_class_apollo.datamodel.RangeI"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left"><B>Fields inherited from interface apollo.datamodel.<A HREF="../../apollo/datamodel/RangeI.html" title="interface in apollo.datamodel">RangeI</A></B></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><A HREF="../../apollo/datamodel/RangeI.html#NO_NAME">NO_NAME</A>, <A HREF="../../apollo/datamodel/RangeI.html#NO_TYPE">NO_TYPE</A></CODE></TD>
</TR>
</TABLE>
&nbsp;
<!-- ======== CONSTRUCTOR SUMMARY ======== -->

<A NAME="constructor_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Constructor Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../apollo/datamodel/Transcript.html#Transcript()">Transcript</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../apollo/datamodel/Transcript.html#Transcript(apollo.datamodel.FeatureSetI)">Transcript</A></B>(<A HREF="../../apollo/datamodel/FeatureSetI.html" title="interface in apollo.datamodel">FeatureSetI</A>&nbsp;fs)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<!-- ========== METHOD SUMMARY =========== -->

<A NAME="method_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Method Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../apollo/datamodel/Transcript.html#addExon(apollo.datamodel.ExonI)">addExon</A></B>(<A HREF="../../apollo/datamodel/ExonI.html" title="interface in apollo.datamodel">ExonI</A>&nbsp;exon)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;If exon overlaps with other exons then it merges them, otherwise
      adds exon to transcript</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../apollo/datamodel/Transcript.html#addFeature(apollo.datamodel.SeqFeatureI)">addFeature</A></B>(<A HREF="../../apollo/datamodel/SeqFeatureI.html" title="interface in apollo.datamodel">SeqFeatureI</A>&nbsp;feature)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Add feature to end of features list, recalc low and high</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../apollo/datamodel/Transcript.html#deleteExon(apollo.datamodel.ExonI)">deleteExon</A></B>(<A HREF="../../apollo/datamodel/ExonI.html" title="interface in apollo.datamodel">ExonI</A>&nbsp;exon)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../apollo/datamodel/Transcript.html#get_cDNA()">get_cDNA</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Calls FeatureSet.get_cDNA which appends exon sequence
      (with getSplicedTranscript) where the actual sequence is retrieved 
      from the ref sequence.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../apollo/datamodel/SequenceI.html" title="interface in apollo.datamodel">SequenceI</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../apollo/datamodel/Transcript.html#get_cDNASequence()">get_cDNASequence</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;by default return null - overridden by Transcript subclass</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../apollo/datamodel/ExonI.html" title="interface in apollo.datamodel">ExonI</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../apollo/datamodel/Transcript.html#getExonAt(int)">getExonAt</A></B>(int&nbsp;i)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.util.Vector</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../apollo/datamodel/Transcript.html#getExons()">getExons</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../apollo/datamodel/AnnotatedFeatureI.html" title="interface in apollo.datamodel">AnnotatedFeatureI</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../apollo/datamodel/Transcript.html#getGene()">getGene</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../apollo/datamodel/Transcript.html#getNonConsensusAcceptorNum()">getNonConsensusAcceptorNum</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;This returns the exon index of the non-consensus acceptor (if
      any), or -1 if there aren't any non-consensus acceptors.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../apollo/datamodel/Transcript.html#getNonConsensusDonorNum()">getNonConsensusDonorNum</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns transcript exon number with non consensus donor.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../apollo/datamodel/SequenceI.html" title="interface in apollo.datamodel">SequenceI</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../apollo/datamodel/Transcript.html#getPeptideSequence()">getPeptideSequence</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;creates peptide sequence if dont have one && prot coding gene - also if pep
 is invalid retranslates</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../apollo/datamodel/Transcript.html#getPeptideValidity()">getPeptideValidity</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../apollo/datamodel/Protein.html" title="class in apollo.datamodel">Protein</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../apollo/datamodel/Transcript.html#getProteinFeat()">getProteinFeat</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;first cut - eventually grow into full fledged protein feat
      that has a pep seq - which will replace Transcripts peptide Seq
      probably will require a Protein class.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../apollo/datamodel/Transcript.html#getTopLevelType()">getTopLevelType</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Transcript has the same biotype as its gene</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../apollo/datamodel/Transcript.html#hasPeptideSequence()">hasPeptideSequence</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;If we have a translation start(nonzero) we have a peptide seq 
      do i have this right? translate may still return null if we have start
      but no transcript sequence(if theres no sequence for the entry).</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../apollo/datamodel/Transcript.html#hasProteinFeat()">hasProteinFeat</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../apollo/datamodel/Transcript.html#haveWholeSequence()">haveWholeSequence</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns true if transcript is hanging off the bounds of the entry
      Does this belong in FeatureSet? Or even SeqFeature?
      Delete this - replaced by isContainedByRefSeq</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../apollo/datamodel/Transcript.html#isProteinCodingGene()">isProteinCodingGene</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;This is the isProteinCodingGene method that really matters--the ones in other classes
      aren't really used.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../apollo/datamodel/Transcript.html#isTranscript()">isTranscript</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns true becuase we have a transcript - its ourself.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../apollo/datamodel/Transcript.html#nonConsensusSplicingOkay()">nonConsensusSplicingOkay</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Return true if non consensus splicing has been deemed ok.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../apollo/datamodel/Transcript.html#nonConsensusSplicingOkay(boolean)">nonConsensusSplicingOkay</A></B>(boolean&nbsp;okay)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Set whether its ok to have non-consensus splice site - no-op overridden
      by Transcript</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../apollo/datamodel/Transcript.html#set_cDNASequence(apollo.datamodel.SequenceI)">set_cDNASequence</A></B>(<A HREF="../../apollo/datamodel/SequenceI.html" title="interface in apollo.datamodel">SequenceI</A>&nbsp;cDNA)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../apollo/datamodel/Transcript.html#setPeptideSequence(apollo.datamodel.SequenceI)">setPeptideSequence</A></B>(<A HREF="../../apollo/datamodel/SequenceI.html" title="interface in apollo.datamodel">SequenceI</A>&nbsp;pep)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../apollo/datamodel/Transcript.html#setPeptideValidity(boolean)">setPeptideValidity</A></B>(boolean&nbsp;validity)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;I think the idea here is if validity is false its saying the peptide is no
 longer valid and a new translation needs to be calculated - this is now
   part of TranslationI - move this method to FeatureSet? (cdna? peptide?)</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../apollo/datamodel/Transcript.html#sortExons()">sortExons</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../apollo/datamodel/Transcript.html#sortTransSpliced()">sortTransSpliced</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../apollo/datamodel/Transcript.html#toString()">toString</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;For debugging</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../apollo/datamodel/Transcript.html#translate()">translate</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Creates SequenceI peptide if doesnt exist already - if translation is null
 or "" peptide will not be created - should it anyways?</TD>
</TR>
</TABLE>
&nbsp;<A NAME="methods_inherited_from_class_apollo.datamodel.AnnotatedFeature"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left"><B>Methods inherited from class apollo.datamodel.<A HREF="../../apollo/datamodel/AnnotatedFeature.html" title="class in apollo.datamodel">AnnotatedFeature</A></B></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><A HREF="../../apollo/datamodel/AnnotatedFeature.html#accept(apollo.util.Visitor)">accept</A>, <A HREF="../../apollo/datamodel/AnnotatedFeature.html#addComment(apollo.datamodel.Comment)">addComment</A>, <A HREF="../../apollo/datamodel/AnnotatedFeature.html#addComment(int, apollo.datamodel.Comment)">addComment</A>, <A HREF="../../apollo/datamodel/AnnotatedFeature.html#addEvidence(apollo.datamodel.Evidence)">addEvidence</A>, <A HREF="../../apollo/datamodel/AnnotatedFeature.html#addEvidence(java.lang.String)">addEvidence</A>, <A HREF="../../apollo/datamodel/AnnotatedFeature.html#addEvidence(java.lang.String, int)">addEvidence</A>, <A HREF="../../apollo/datamodel/AnnotatedFeature.html#addEvidence(java.lang.String, java.lang.String, int)">addEvidence</A>, <A HREF="../../apollo/datamodel/AnnotatedFeature.html#addSynonym(int, java.lang.String)">addSynonym</A>, <A HREF="../../apollo/datamodel/AnnotatedFeature.html#addSynonym(java.lang.String)">addSynonym</A>, <A HREF="../../apollo/datamodel/AnnotatedFeature.html#addSynonym(apollo.datamodel.Synonym)">addSynonym</A>, <A HREF="../../apollo/datamodel/AnnotatedFeature.html#clearComments()">clearComments</A>, <A HREF="../../apollo/datamodel/AnnotatedFeature.html#clearSynonyms()">clearSynonyms</A>, <A HREF="../../apollo/datamodel/AnnotatedFeature.html#clone()">clone</A>, <A HREF="../../apollo/datamodel/AnnotatedFeature.html#cloneAnnot()">cloneAnnot</A>, <A HREF="../../apollo/datamodel/AnnotatedFeature.html#deleteComment(apollo.datamodel.Comment)">deleteComment</A>, <A HREF="../../apollo/datamodel/AnnotatedFeature.html#deleteComment(int)">deleteComment</A>, <A HREF="../../apollo/datamodel/AnnotatedFeature.html#deleteEvidence(java.lang.String)">deleteEvidence</A>, <A HREF="../../apollo/datamodel/AnnotatedFeature.html#deleteSynonym(java.lang.String)">deleteSynonym</A>, <A HREF="../../apollo/datamodel/AnnotatedFeature.html#deleteSynonym(apollo.datamodel.Synonym)">deleteSynonym</A>, <A HREF="../../apollo/datamodel/AnnotatedFeature.html#getAnnotatedFeature()">getAnnotatedFeature</A>, <A HREF="../../apollo/datamodel/AnnotatedFeature.html#getAnnotChild(int)">getAnnotChild</A>, <A HREF="../../apollo/datamodel/AnnotatedFeature.html#getCommentIndex(apollo.datamodel.Comment)">getCommentIndex</A>, <A HREF="../../apollo/datamodel/AnnotatedFeature.html#getComments()">getComments</A>, <A HREF="../../apollo/datamodel/AnnotatedFeature.html#getDescription()">getDescription</A>, <A HREF="../../apollo/datamodel/AnnotatedFeature.html#getEvidence()">getEvidence</A>, <A HREF="../../apollo/datamodel/AnnotatedFeature.html#getEvidenceFinder()">getEvidenceFinder</A>, <A HREF="../../apollo/datamodel/AnnotatedFeature.html#getOwner()">getOwner</A>, <A HREF="../../apollo/datamodel/AnnotatedFeature.html#getSynonym(int)">getSynonym</A>, <A HREF="../../apollo/datamodel/AnnotatedFeature.html#getSynonyms()">getSynonyms</A>, <A HREF="../../apollo/datamodel/AnnotatedFeature.html#getSynonyms(boolean)">getSynonyms</A>, <A HREF="../../apollo/datamodel/AnnotatedFeature.html#getSynonymSize()">getSynonymSize</A>, <A HREF="../../apollo/datamodel/AnnotatedFeature.html#hasAnnotatedFeature()">hasAnnotatedFeature</A>, <A HREF="../../apollo/datamodel/AnnotatedFeature.html#hasSynonym(java.lang.String)">hasSynonym</A>, <A HREF="../../apollo/datamodel/AnnotatedFeature.html#isAnnotTop()">isAnnotTop</A>, <A HREF="../../apollo/datamodel/AnnotatedFeature.html#isFinished()">isFinished</A>, <A HREF="../../apollo/datamodel/AnnotatedFeature.html#isProblematic()">isProblematic</A>, <A HREF="../../apollo/datamodel/AnnotatedFeature.html#setDescription(java.lang.String)">setDescription</A>, <A HREF="../../apollo/datamodel/AnnotatedFeature.html#setEvidenceFinder(apollo.datamodel.EvidenceFinder)">setEvidenceFinder</A>, <A HREF="../../apollo/datamodel/AnnotatedFeature.html#setIsProblematic(boolean)">setIsProblematic</A>, <A HREF="../../apollo/datamodel/AnnotatedFeature.html#setOwner(java.lang.String)">setOwner</A></CODE></TD>
</TR>
</TABLE>
&nbsp;<A NAME="methods_inherited_from_class_apollo.datamodel.FeatureSet"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left"><B>Methods inherited from class apollo.datamodel.<A HREF="../../apollo/datamodel/FeatureSet.html" title="class in apollo.datamodel">FeatureSet</A></B></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><A HREF="../../apollo/datamodel/FeatureSet.html#addFeature(apollo.datamodel.SeqFeatureI, boolean)">addFeature</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#adjustEdges()">adjustEdges</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#adjustEdges(apollo.datamodel.SeqFeatureI)">adjustEdges</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#beforeFivePrimeEnd(apollo.datamodel.SeqFeatureI)">beforeFivePrimeEnd</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#buildEditList()">buildEditList</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#buildORFEditList()">buildORFEditList</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#calcTranslationStartForLongestPeptide()">calcTranslationStartForLongestPeptide</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#canHaveChildren()">canHaveChildren</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#clearKids()">clearKids</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#deleteFeature(apollo.datamodel.SeqFeatureI)">deleteFeature</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#deleteFeatureAt(int)">deleteFeatureAt</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#findFeaturesByAllNames(java.lang.String)">findFeaturesByAllNames</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#findFeaturesByAllNames(java.lang.String, boolean)">findFeaturesByAllNames</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#findFeaturesByAllNames(java.lang.String, boolean, boolean)">findFeaturesByAllNames</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#findFeaturesByHitName(java.lang.String)">findFeaturesByHitName</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#findFeaturesByName(java.lang.String)">findFeaturesByName</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#findFeaturesByName(java.lang.String, boolean)">findFeaturesByName</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#flipFlop()">flipFlop</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#get_ORF(java.lang.String)">get_ORF</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#get_ORF(java.lang.String, int, int)">get_ORF</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#getFeatureAt(int)">getFeatureAt</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#getFeatureContaining(int)">getFeatureContaining</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#getFeatureIndex(apollo.datamodel.SeqFeatureI)">getFeatureIndex</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#getFeaturePosition(int)">getFeaturePosition</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#getFeatures()">getFeatures</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#getGenomicPosForPeptidePos(int)">getGenomicPosForPeptidePos</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#getGenomicPosition(int)">getGenomicPosition</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#getHitSequence()">getHitSequence</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#getIndexContaining(int)">getIndexContaining</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#getLastBaseOfStopCodon()">getLastBaseOfStopCodon</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#getLeafFeatsOver(int)">getLeafFeatsOver</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#getName()">getName</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#getNumberOfDescendents()">getNumberOfDescendents</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#getPeptidePosForGenomicPos(int)">getPeptidePosForGenomicPos</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#getPositionFrom(int, int)">getPositionFrom</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#getScore()">getScore</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#getSequencingErrorAtPosition(int)">getSequencingErrorAtPosition</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#getSplicedLength()">getSplicedLength</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#getSplicedLength(int, int)">getSplicedLength</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#getSplicedTranscript(int, int)">getSplicedTranscript</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#getStartAA()">getStartAA</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#getStartCodon()">getStartCodon</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#getTranslation()">getTranslation</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#getTranslationEnd()">getTranslationEnd</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#getTranslationRange()">getTranslationRange</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#getTranslationStart()">getTranslationStart</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#hasDescendents()">hasDescendents</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#hasNameBeenSet()">hasNameBeenSet</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#hasReadThroughStop()">hasReadThroughStop</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#hasTranslation()">hasTranslation</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#hasTranslationEnd()">hasTranslationEnd</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#hasTranslationStart()">hasTranslationStart</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#insertFeatureAt(apollo.datamodel.SeqFeatureI, int)">insertFeatureAt</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#isFlagSet(int)">isFlagSet</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#isMissing3prime()">isMissing3prime</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#isMissing5prime()">isMissing5prime</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#isSequencingErrorPosition(int)">isSequencingErrorPosition</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#isTransSpliced()">isTransSpliced</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#minus1FrameShiftPosition()">minus1FrameShiftPosition</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#pastThreePrimeEnd(apollo.datamodel.SeqFeatureI)">pastThreePrimeEnd</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#plus1FrameShiftPosition()">plus1FrameShiftPosition</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#rangeIsUnassigned()">rangeIsUnassigned</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#readThroughStopPosition()">readThroughStopPosition</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#readThroughStopResidue()">readThroughStopResidue</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#setFlag(boolean, byte)">setFlag</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#setHitSequence(apollo.datamodel.SequenceI)">setHitSequence</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#setMinus1FrameShiftPosition(int)">setMinus1FrameShiftPosition</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#setMissing3prime(boolean)">setMissing3prime</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#setMissing5prime(boolean)">setMissing5prime</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#setPhases()">setPhases</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#setPlus1FrameShiftPosition(int)">setPlus1FrameShiftPosition</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#setProteinCodingGene(boolean)">setProteinCodingGene</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#setReadThroughStop(boolean)">setReadThroughStop</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#setReadThroughStop(java.lang.String)">setReadThroughStop</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#setRefSequence(apollo.datamodel.SequenceI)">setRefSequence</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#setTranslationEnd(int)">setTranslationEnd</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#setTranslationEndFromStart()">setTranslationEndFromStart</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#setTranslationEndNoPhase(int)">setTranslationEndNoPhase</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#setTranslationStart(int)">setTranslationStart</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#setTranslationStart(int, boolean)">setTranslationStart</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#size()">size</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#sort(int)">sort</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#sort(int, boolean)">sort</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#unConventionalStart()">unConventionalStart</A>, <A HREF="../../apollo/datamodel/FeatureSet.html#withinCDS(int)">withinCDS</A></CODE></TD>
</TR>
</TABLE>
&nbsp;<A NAME="methods_inherited_from_class_apollo.datamodel.SeqFeature"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left"><B>Methods inherited from class apollo.datamodel.<A HREF="../../apollo/datamodel/SeqFeature.html" title="class in apollo.datamodel">SeqFeature</A></B></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><A HREF="../../apollo/datamodel/SeqFeature.html#addDbXref(apollo.datamodel.DbXref)">addDbXref</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#addProperty(java.lang.String, java.lang.String)">addProperty</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#addRefFeature(java.lang.String, apollo.datamodel.SeqFeatureI)">addRefFeature</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#addScore(double)">addScore</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#addScore(apollo.datamodel.Score)">addScore</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#addScore(java.lang.String, double)">addScore</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#addScore(java.lang.String, java.lang.String)">addScore</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#alignmentIsPeptide()">alignmentIsPeptide</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#amend_RNA(java.lang.StringBuffer)">amend_RNA</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#buildmRNAEditList()">buildmRNAEditList</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#clearProperties()">clearProperties</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#cloneFeature()">cloneFeature</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#compareTo(java.lang.Object)">compareTo</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#descendsFrom(apollo.datamodel.SeqFeatureI)">descendsFrom</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#getAlignment()">getAlignment</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#getAnalogousOppositeStrandFeature()">getAnalogousOppositeStrandFeature</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#getCigar()">getCigar</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#getCloneSource()">getCloneSource</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#getCodingDNA()">getCodingDNA</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#getCodingProperties()">getCodingProperties</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#getDatabase()">getDatabase</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#getDbXref(int)">getDbXref</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#getDbXrefs()">getDbXrefs</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#getEndPhase()">getEndPhase</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#getExplicitAlignment()">getExplicitAlignment</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#getFeatureSequence()">getFeatureSequence</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#getFrame()">getFrame</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#getGenomicErrors()">getGenomicErrors</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#getHend()">getHend</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#getHhigh()">getHhigh</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#getHitFeature()">getHitFeature</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#getHlow()">getHlow</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#getHname()">getHname</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#getHstart()">getHstart</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#getHstrand()">getHstrand</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#getId()">getId</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#getIdentifier()">getIdentifier</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#getNumberOfChildren()">getNumberOfChildren</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#getParent()">getParent</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#getPhase()">getPhase</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#getPrimaryDbXref()">getPrimaryDbXref</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#getProgramName()">getProgramName</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#getProperties()">getProperties</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#getPropertiesMulti()">getPropertiesMulti</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#getProperty(java.lang.String)">getProperty</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#getPropertyMulti(java.lang.String)">getPropertyMulti</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#getRefFeature()">getRefFeature</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#getRefFeature(java.lang.String)">getRefFeature</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#getRefId()">getRefId</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#getScore(java.lang.String)">getScore</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#getScores()">getScores</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#getStrandedFeatSetAncestor()">getStrandedFeatSetAncestor</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#getSubSequence(java.lang.String, int, int)">getSubSequence</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#getSyntenyLinkInfo()">getSyntenyLinkInfo</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#getUnpaddedAlignment()">getUnpaddedAlignment</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#getUserObject()">getUserObject</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#hasAlignable()">hasAlignable</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#hasAnalogousOppositeStrandFeature()">hasAnalogousOppositeStrandFeature</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#hasCigar()">hasCigar</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#hasHitFeature()">hasHitFeature</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#hasId()">hasId</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#hasKids()">hasKids</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#hasSyntenyLinkInfo()">hasSyntenyLinkInfo</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#haveExplicitAlignment()">haveExplicitAlignment</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#haveRealAlignment()">haveRealAlignment</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#initWithSeqFeat(apollo.datamodel.SeqFeatureI)">initWithSeqFeat</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#isAncestorOf(apollo.datamodel.SeqFeatureI)">isAncestorOf</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#isAnnot()">isAnnot</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#isClone()">isClone</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#isCodon()">isCodon</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#isExon()">isExon</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#isProtein()">isProtein</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#isSameFeat(apollo.datamodel.SeqFeatureI)">isSameFeat</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#isSequencingError()">isSequencingError</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#main(java.lang.String[])">main</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#merge(apollo.datamodel.SeqFeatureI)">merge</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#na2aa(java.lang.String)">na2aa</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#numberOfGenerations()">numberOfGenerations</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#parseCigar()">parseCigar</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#removeProperty(java.lang.String)">removeProperty</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#replaceProperty(java.lang.String, java.lang.String)">replaceProperty</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#setAlignment(java.lang.String)">setAlignment</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#setAnalogousOppositeStrandFeature(apollo.datamodel.SeqFeatureI)">setAnalogousOppositeStrandFeature</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#setCigar(java.lang.String)">setCigar</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#setDatabase(java.lang.String)">setDatabase</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#setExplicitAlignment(java.lang.String)">setExplicitAlignment</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#setId(java.lang.String)">setId</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#setIdentifier(apollo.datamodel.Identifier)">setIdentifier</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#setPhase(int)">setPhase</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#setProgramName(java.lang.String)">setProgramName</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#setQueryFeature(apollo.datamodel.SeqFeatureI)">setQueryFeature</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#setRefFeature(apollo.datamodel.SeqFeatureI)">setRefFeature</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#setScore(double)">setScore</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#setSyntenyLinkInfo(java.lang.String)">setSyntenyLinkInfo</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#setTopLevelType(java.lang.String)">setTopLevelType</A>, <A HREF="../../apollo/datamodel/SeqFeature.html#setUserObject(java.lang.Object)">setUserObject</A></CODE></TD>
</TR>
</TABLE>
&nbsp;<A NAME="methods_inherited_from_class_apollo.datamodel.Range"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left"><B>Methods inherited from class apollo.datamodel.<A HREF="../../apollo/datamodel/Range.html" title="class in apollo.datamodel">Range</A></B></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><A HREF="../../apollo/datamodel/Range.html#contains(int)">contains</A>, <A HREF="../../apollo/datamodel/Range.html#contains(apollo.datamodel.RangeI)">contains</A>, <A HREF="../../apollo/datamodel/Range.html#convertFromBaseOrientedToInterbase()">convertFromBaseOrientedToInterbase</A>, <A HREF="../../apollo/datamodel/Range.html#convertFromInterbaseToBaseOriented()">convertFromInterbaseToBaseOriented</A>, <A HREF="../../apollo/datamodel/Range.html#getEnd()">getEnd</A>, <A HREF="../../apollo/datamodel/Range.html#getEndAsString()">getEndAsString</A>, <A HREF="../../apollo/datamodel/Range.html#getFeatureType()">getFeatureType</A>, <A HREF="../../apollo/datamodel/Range.html#getHigh()">getHigh</A>, <A HREF="../../apollo/datamodel/Range.html#getLeftOverlap(apollo.datamodel.RangeI)">getLeftOverlap</A>, <A HREF="../../apollo/datamodel/Range.html#getLow()">getLow</A>, <A HREF="../../apollo/datamodel/Range.html#getRangeClone()">getRangeClone</A>, <A HREF="../../apollo/datamodel/Range.html#getRefSequence()">getRefSequence</A>, <A HREF="../../apollo/datamodel/Range.html#getResidues()">getResidues</A>, <A HREF="../../apollo/datamodel/Range.html#getRightOverlap(apollo.datamodel.RangeI)">getRightOverlap</A>, <A HREF="../../apollo/datamodel/Range.html#getStart()">getStart</A>, <A HREF="../../apollo/datamodel/Range.html#getStartAsString()">getStartAsString</A>, <A HREF="../../apollo/datamodel/Range.html#getStrand()">getStrand</A>, <A HREF="../../apollo/datamodel/Range.html#hasFeatureType()">hasFeatureType</A>, <A HREF="../../apollo/datamodel/Range.html#hasName()">hasName</A>, <A HREF="../../apollo/datamodel/Range.html#hasRefSequence()">hasRefSequence</A>, <A HREF="../../apollo/datamodel/Range.html#isContainedByRefSeq()">isContainedByRefSeq</A>, <A HREF="../../apollo/datamodel/Range.html#isExactOverlap(apollo.datamodel.RangeI)">isExactOverlap</A>, <A HREF="../../apollo/datamodel/Range.html#isForwardStrand()">isForwardStrand</A>, <A HREF="../../apollo/datamodel/Range.html#isIdentical(apollo.datamodel.RangeI)">isIdentical</A>, <A HREF="../../apollo/datamodel/Range.html#isSequenceAvailable(int)">isSequenceAvailable</A>, <A HREF="../../apollo/datamodel/Range.html#length()">length</A>, <A HREF="../../apollo/datamodel/Range.html#overlaps(apollo.datamodel.RangeI)">overlaps</A>, <A HREF="../../apollo/datamodel/Range.html#sameRange(apollo.datamodel.RangeI)">sameRange</A>, <A HREF="../../apollo/datamodel/Range.html#setEnd(int)">setEnd</A>, <A HREF="../../apollo/datamodel/Range.html#setFeatureType(java.lang.String)">setFeatureType</A>, <A HREF="../../apollo/datamodel/Range.html#setHigh(int)">setHigh</A>, <A HREF="../../apollo/datamodel/Range.html#setLow(int)">setLow</A>, <A HREF="../../apollo/datamodel/Range.html#setName(java.lang.String)">setName</A>, <A HREF="../../apollo/datamodel/Range.html#setStart(int)">setStart</A>, <A HREF="../../apollo/datamodel/Range.html#setStrand(int)">setStrand</A></CODE></TD>
</TR>
</TABLE>
&nbsp;<A NAME="methods_inherited_from_class_java.lang.Object"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left"><B>Methods inherited from class java.lang.Object</B></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE>equals, finalize, getClass, hashCode, notify, notifyAll, wait, wait, wait</CODE></TD>
</TR>
</TABLE>
&nbsp;<A NAME="methods_inherited_from_class_apollo.datamodel.AnnotatedFeatureI"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left"><B>Methods inherited from interface apollo.datamodel.<A HREF="../../apollo/datamodel/AnnotatedFeatureI.html" title="interface in apollo.datamodel">AnnotatedFeatureI</A></B></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><A HREF="../../apollo/datamodel/AnnotatedFeatureI.html#getIdentifier()">getIdentifier</A>, <A HREF="../../apollo/datamodel/AnnotatedFeatureI.html#setIdentifier(apollo.datamodel.Identifier)">setIdentifier</A></CODE></TD>
</TR>
</TABLE>
&nbsp;<A NAME="methods_inherited_from_class_apollo.datamodel.FeatureSetI"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left"><B>Methods inherited from interface apollo.datamodel.<A HREF="../../apollo/datamodel/FeatureSetI.html" title="interface in apollo.datamodel">FeatureSetI</A></B></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><A HREF="../../apollo/datamodel/FeatureSetI.html#adjustEdges()">adjustEdges</A>, <A HREF="../../apollo/datamodel/FeatureSetI.html#adjustEdges(apollo.datamodel.SeqFeatureI)">adjustEdges</A>, <A HREF="../../apollo/datamodel/FeatureSetI.html#buildEditList()">buildEditList</A>, <A HREF="../../apollo/datamodel/FeatureSetI.html#deleteFeature(apollo.datamodel.SeqFeatureI)">deleteFeature</A>, <A HREF="../../apollo/datamodel/FeatureSetI.html#deleteFeatureAt(int)">deleteFeatureAt</A>, <A HREF="../../apollo/datamodel/FeatureSetI.html#findFeaturesByAllNames(java.lang.String)">findFeaturesByAllNames</A>, <A HREF="../../apollo/datamodel/FeatureSetI.html#findFeaturesByAllNames(java.lang.String, boolean)">findFeaturesByAllNames</A>, <A HREF="../../apollo/datamodel/FeatureSetI.html#findFeaturesByAllNames(java.lang.String, boolean, boolean)">findFeaturesByAllNames</A>, <A HREF="../../apollo/datamodel/FeatureSetI.html#findFeaturesByHitName(java.lang.String)">findFeaturesByHitName</A>, <A HREF="../../apollo/datamodel/FeatureSetI.html#findFeaturesByName(java.lang.String)">findFeaturesByName</A>, <A HREF="../../apollo/datamodel/FeatureSetI.html#findFeaturesByName(java.lang.String, boolean)">findFeaturesByName</A>, <A HREF="../../apollo/datamodel/FeatureSetI.html#getHitSequence()">getHitSequence</A>, <A HREF="../../apollo/datamodel/FeatureSetI.html#getIndexContaining(int)">getIndexContaining</A>, <A HREF="../../apollo/datamodel/FeatureSetI.html#getPositionFrom(int, int)">getPositionFrom</A>, <A HREF="../../apollo/datamodel/FeatureSetI.html#getSequencingErrorAtPosition(int)">getSequencingErrorAtPosition</A>, <A HREF="../../apollo/datamodel/FeatureSetI.html#getSplicedLength()">getSplicedLength</A>, <A HREF="../../apollo/datamodel/FeatureSetI.html#getStartAA()">getStartAA</A>, <A HREF="../../apollo/datamodel/FeatureSetI.html#getStartCodon()">getStartCodon</A>, <A HREF="../../apollo/datamodel/FeatureSetI.html#hasDescendents()">hasDescendents</A>, <A HREF="../../apollo/datamodel/FeatureSetI.html#hasNameBeenSet()">hasNameBeenSet</A>, <A HREF="../../apollo/datamodel/FeatureSetI.html#hasReadThroughStop()">hasReadThroughStop</A>, <A HREF="../../apollo/datamodel/FeatureSetI.html#isFlagSet(int)">isFlagSet</A>, <A HREF="../../apollo/datamodel/FeatureSetI.html#isSequencingErrorPosition(int)">isSequencingErrorPosition</A>, <A HREF="../../apollo/datamodel/FeatureSetI.html#minus1FrameShiftPosition()">minus1FrameShiftPosition</A>, <A HREF="../../apollo/datamodel/FeatureSetI.html#plus1FrameShiftPosition()">plus1FrameShiftPosition</A>, <A HREF="../../apollo/datamodel/FeatureSetI.html#readThroughStopPosition()">readThroughStopPosition</A>, <A HREF="../../apollo/datamodel/FeatureSetI.html#readThroughStopResidue()">readThroughStopResidue</A>, <A HREF="../../apollo/datamodel/FeatureSetI.html#setFlag(boolean, byte)">setFlag</A>, <A HREF="../../apollo/datamodel/FeatureSetI.html#setHitSequence(apollo.datamodel.SequenceI)">setHitSequence</A>, <A HREF="../../apollo/datamodel/FeatureSetI.html#setMinus1FrameShiftPosition(int)">setMinus1FrameShiftPosition</A>, <A HREF="../../apollo/datamodel/FeatureSetI.html#setPlus1FrameShiftPosition(int)">setPlus1FrameShiftPosition</A>, <A HREF="../../apollo/datamodel/FeatureSetI.html#setProteinCodingGene(boolean)">setProteinCodingGene</A>, <A HREF="../../apollo/datamodel/FeatureSetI.html#setReadThroughStop(boolean)">setReadThroughStop</A>, <A HREF="../../apollo/datamodel/FeatureSetI.html#setReadThroughStop(java.lang.String)">setReadThroughStop</A>, <A HREF="../../apollo/datamodel/FeatureSetI.html#sort(int)">sort</A>, <A HREF="../../apollo/datamodel/FeatureSetI.html#sort(int, boolean)">sort</A>, <A HREF="../../apollo/datamodel/FeatureSetI.html#unConventionalStart()">unConventionalStart</A>, <A HREF="../../apollo/datamodel/FeatureSetI.html#withinCDS(int)">withinCDS</A></CODE></TD>
</TR>
</TABLE>
&nbsp;<A NAME="methods_inherited_from_class_apollo.datamodel.SeqFeatureI"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left"><B>Methods inherited from interface apollo.datamodel.<A HREF="../../apollo/datamodel/SeqFeatureI.html" title="interface in apollo.datamodel">SeqFeatureI</A></B></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><A HREF="../../apollo/datamodel/SeqFeatureI.html#addDbXref(apollo.datamodel.DbXref)">addDbXref</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#addFeature(apollo.datamodel.SeqFeatureI, boolean)">addFeature</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#addProperty(java.lang.String, java.lang.String)">addProperty</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#addScore(double)">addScore</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#addScore(apollo.datamodel.Score)">addScore</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#addScore(java.lang.String, double)">addScore</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#addScore(java.lang.String, java.lang.String)">addScore</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#alignmentIsPeptide()">alignmentIsPeptide</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#clearKids()">clearKids</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#clearProperties()">clearProperties</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#cloneFeature()">cloneFeature</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#compareTo(java.lang.Object)">compareTo</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#descendsFrom(apollo.datamodel.SeqFeatureI)">descendsFrom</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#flipFlop()">flipFlop</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#getAlignment()">getAlignment</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#getAnalogousOppositeStrandFeature()">getAnalogousOppositeStrandFeature</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#getCigar()">getCigar</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#getCloneSource()">getCloneSource</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#getCodingDNA()">getCodingDNA</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#getCodingProperties()">getCodingProperties</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#getDatabase()">getDatabase</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#getDbXref(int)">getDbXref</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#getDbXrefs()">getDbXrefs</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#getEndPhase()">getEndPhase</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#getExplicitAlignment()">getExplicitAlignment</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#getFeatureAt(int)">getFeatureAt</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#getFeatureContaining(int)">getFeatureContaining</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#getFeatureIndex(apollo.datamodel.SeqFeatureI)">getFeatureIndex</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#getFeaturePosition(int)">getFeaturePosition</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#getFeatures()">getFeatures</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#getFeatureSequence()">getFeatureSequence</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#getFrame()">getFrame</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#getGenomicErrors()">getGenomicErrors</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#getGenomicPosForPeptidePos(int)">getGenomicPosForPeptidePos</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#getGenomicPosition(int)">getGenomicPosition</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#getHend()">getHend</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#getHhigh()">getHhigh</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#getHitFeature()">getHitFeature</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#getHlow()">getHlow</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#getHname()">getHname</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#getHstart()">getHstart</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#getHstrand()">getHstrand</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#getId()">getId</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#getLeafFeatsOver(int)">getLeafFeatsOver</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#getNumberOfChildren()">getNumberOfChildren</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#getNumberOfDescendents()">getNumberOfDescendents</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#getParent()">getParent</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#getPhase()">getPhase</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#getProgramName()">getProgramName</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#getProperties()">getProperties</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#getPropertiesMulti()">getPropertiesMulti</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#getProperty(java.lang.String)">getProperty</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#getPropertyMulti(java.lang.String)">getPropertyMulti</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#getRefFeature()">getRefFeature</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#getRefId()">getRefId</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#getScore()">getScore</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#getScore(java.lang.String)">getScore</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#getScores()">getScores</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#getStrandedFeatSetAncestor()">getStrandedFeatSetAncestor</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#getSyntenyLinkInfo()">getSyntenyLinkInfo</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#getTranslation()">getTranslation</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#getUnpaddedAlignment()">getUnpaddedAlignment</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#getUserObject()">getUserObject</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#hasAlignable()">hasAlignable</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#hasAnalogousOppositeStrandFeature()">hasAnalogousOppositeStrandFeature</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#hasHitFeature()">hasHitFeature</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#hasId()">hasId</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#hasKids()">hasKids</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#hasSyntenyLinkInfo()">hasSyntenyLinkInfo</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#hasTranslation()">hasTranslation</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#haveExplicitAlignment()">haveExplicitAlignment</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#haveRealAlignment()">haveRealAlignment</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#isAncestorOf(apollo.datamodel.SeqFeatureI)">isAncestorOf</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#isAnnot()">isAnnot</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#isClone()">isClone</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#isCodon()">isCodon</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#isExon()">isExon</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#isProtein()">isProtein</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#isSameFeat(apollo.datamodel.SeqFeatureI)">isSameFeat</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#isSequencingError()">isSequencingError</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#merge(apollo.datamodel.SeqFeatureI)">merge</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#numberOfGenerations()">numberOfGenerations</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#parseCigar()">parseCigar</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#removeProperty(java.lang.String)">removeProperty</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#replaceProperty(java.lang.String, java.lang.String)">replaceProperty</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#setAlignment(java.lang.String)">setAlignment</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#setAnalogousOppositeStrandFeature(apollo.datamodel.SeqFeatureI)">setAnalogousOppositeStrandFeature</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#setCigar(java.lang.String)">setCigar</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#setDatabase(java.lang.String)">setDatabase</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#setExplicitAlignment(java.lang.String)">setExplicitAlignment</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#setId(java.lang.String)">setId</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#setPhase(int)">setPhase</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#setProgramName(java.lang.String)">setProgramName</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#setQueryFeature(apollo.datamodel.SeqFeatureI)">setQueryFeature</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#setRefFeature(apollo.datamodel.SeqFeatureI)">setRefFeature</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#setScore(double)">setScore</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#setSyntenyLinkInfo(java.lang.String)">setSyntenyLinkInfo</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#setTopLevelType(java.lang.String)">setTopLevelType</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#setUserObject(java.lang.Object)">setUserObject</A>, <A HREF="../../apollo/datamodel/SeqFeatureI.html#size()">size</A></CODE></TD>
</TR>
</TABLE>
&nbsp;<A NAME="methods_inherited_from_class_apollo.datamodel.RangeI"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left"><B>Methods inherited from interface apollo.datamodel.<A HREF="../../apollo/datamodel/RangeI.html" title="interface in apollo.datamodel">RangeI</A></B></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><A HREF="../../apollo/datamodel/RangeI.html#canHaveChildren()">canHaveChildren</A>, <A HREF="../../apollo/datamodel/RangeI.html#contains(int)">contains</A>, <A HREF="../../apollo/datamodel/RangeI.html#contains(apollo.datamodel.RangeI)">contains</A>, <A HREF="../../apollo/datamodel/RangeI.html#convertFromBaseOrientedToInterbase()">convertFromBaseOrientedToInterbase</A>, <A HREF="../../apollo/datamodel/RangeI.html#convertFromInterbaseToBaseOriented()">convertFromInterbaseToBaseOriented</A>, <A HREF="../../apollo/datamodel/RangeI.html#getEnd()">getEnd</A>, <A HREF="../../apollo/datamodel/RangeI.html#getFeatureType()">getFeatureType</A>, <A HREF="../../apollo/datamodel/RangeI.html#getHigh()">getHigh</A>, <A HREF="../../apollo/datamodel/RangeI.html#getLeftOverlap(apollo.datamodel.RangeI)">getLeftOverlap</A>, <A HREF="../../apollo/datamodel/RangeI.html#getLow()">getLow</A>, <A HREF="../../apollo/datamodel/RangeI.html#getName()">getName</A>, <A HREF="../../apollo/datamodel/RangeI.html#getRangeClone()">getRangeClone</A>, <A HREF="../../apollo/datamodel/RangeI.html#getRefSequence()">getRefSequence</A>, <A HREF="../../apollo/datamodel/RangeI.html#getResidues()">getResidues</A>, <A HREF="../../apollo/datamodel/RangeI.html#getRightOverlap(apollo.datamodel.RangeI)">getRightOverlap</A>, <A HREF="../../apollo/datamodel/RangeI.html#getStart()">getStart</A>, <A HREF="../../apollo/datamodel/RangeI.html#getStrand()">getStrand</A>, <A HREF="../../apollo/datamodel/RangeI.html#hasFeatureType()">hasFeatureType</A>, <A HREF="../../apollo/datamodel/RangeI.html#hasName()">hasName</A>, <A HREF="../../apollo/datamodel/RangeI.html#hasRefSequence()">hasRefSequence</A>, <A HREF="../../apollo/datamodel/RangeI.html#isContainedByRefSeq()">isContainedByRefSeq</A>, <A HREF="../../apollo/datamodel/RangeI.html#isExactOverlap(apollo.datamodel.RangeI)">isExactOverlap</A>, <A HREF="../../apollo/datamodel/RangeI.html#isForwardStrand()">isForwardStrand</A>, <A HREF="../../apollo/datamodel/RangeI.html#isIdentical(apollo.datamodel.RangeI)">isIdentical</A>, <A HREF="../../apollo/datamodel/RangeI.html#isSequenceAvailable(int)">isSequenceAvailable</A>, <A HREF="../../apollo/datamodel/RangeI.html#length()">length</A>, <A HREF="../../apollo/datamodel/RangeI.html#overlaps(apollo.datamodel.RangeI)">overlaps</A>, <A HREF="../../apollo/datamodel/RangeI.html#rangeIsUnassigned()">rangeIsUnassigned</A>, <A HREF="../../apollo/datamodel/RangeI.html#sameRange(apollo.datamodel.RangeI)">sameRange</A>, <A HREF="../../apollo/datamodel/RangeI.html#setEnd(int)">setEnd</A>, <A HREF="../../apollo/datamodel/RangeI.html#setFeatureType(java.lang.String)">setFeatureType</A>, <A HREF="../../apollo/datamodel/RangeI.html#setHigh(int)">setHigh</A>, <A HREF="../../apollo/datamodel/RangeI.html#setLow(int)">setLow</A>, <A HREF="../../apollo/datamodel/RangeI.html#setName(java.lang.String)">setName</A>, <A HREF="../../apollo/datamodel/RangeI.html#setRefSequence(apollo.datamodel.SequenceI)">setRefSequence</A>, <A HREF="../../apollo/datamodel/RangeI.html#setStart(int)">setStart</A>, <A HREF="../../apollo/datamodel/RangeI.html#setStrand(int)">setStrand</A></CODE></TD>
</TR>
</TABLE>
&nbsp;<A NAME="methods_inherited_from_class_apollo.datamodel.TranslationI"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left"><B>Methods inherited from interface apollo.datamodel.<A HREF="../../apollo/datamodel/TranslationI.html" title="interface in apollo.datamodel">TranslationI</A></B></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><A HREF="../../apollo/datamodel/TranslationI.html#calcTranslationStartForLongestPeptide()">calcTranslationStartForLongestPeptide</A>, <A HREF="../../apollo/datamodel/TranslationI.html#getLastBaseOfStopCodon()">getLastBaseOfStopCodon</A>, <A HREF="../../apollo/datamodel/TranslationI.html#getTranslationEnd()">getTranslationEnd</A>, <A HREF="../../apollo/datamodel/TranslationI.html#getTranslationRange()">getTranslationRange</A>, <A HREF="../../apollo/datamodel/TranslationI.html#getTranslationStart()">getTranslationStart</A>, <A HREF="../../apollo/datamodel/TranslationI.html#hasTranslationEnd()">hasTranslationEnd</A>, <A HREF="../../apollo/datamodel/TranslationI.html#hasTranslationStart()">hasTranslationStart</A>, <A HREF="../../apollo/datamodel/TranslationI.html#isMissing3prime()">isMissing3prime</A>, <A HREF="../../apollo/datamodel/TranslationI.html#isMissing5prime()">isMissing5prime</A>, <A HREF="../../apollo/datamodel/TranslationI.html#setMissing3prime(boolean)">setMissing3prime</A>, <A HREF="../../apollo/datamodel/TranslationI.html#setMissing5prime(boolean)">setMissing5prime</A>, <A HREF="../../apollo/datamodel/TranslationI.html#setTranslationEnd(int)">setTranslationEnd</A>, <A HREF="../../apollo/datamodel/TranslationI.html#setTranslationEndFromStart()">setTranslationEndFromStart</A>, <A HREF="../../apollo/datamodel/TranslationI.html#setTranslationStart(int)">setTranslationStart</A>, <A HREF="../../apollo/datamodel/TranslationI.html#setTranslationStart(int, boolean)">setTranslationStart</A></CODE></TD>
</TR>
</TABLE>
&nbsp;
<P>

<!-- ============ FIELD DETAIL =========== -->

<A NAME="field_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Field Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="logger"><!-- --></A><H3>
logger</H3>
<PRE>
protected static final org.apache.log4j.Logger <B>logger</B></PRE>
<DL>
<DL>
</DL>
</DL>
<HR>

<A NAME="TRANSCRIPT_TYPE"><!-- --></A><H3>
TRANSCRIPT_TYPE</H3>
<PRE>
public static final java.lang.String <B>TRANSCRIPT_TYPE</B></PRE>
<DL>
<DL>
<DT><B>See Also:</B><DD><A HREF="../../constant-values.html#apollo.datamodel.Transcript.TRANSCRIPT_TYPE">Constant Field Values</A></DL>
</DL>
<HR>

<A NAME="cDNA"><!-- --></A><H3>
cDNA</H3>
<PRE>
protected <A HREF="../../apollo/datamodel/SequenceI.html" title="interface in apollo.datamodel">SequenceI</A> <B>cDNA</B></PRE>
<DL>
<DL>
</DL>
</DL>
<HR>

<A NAME="peptide"><!-- --></A><H3>
peptide</H3>
<PRE>
protected <A HREF="../../apollo/datamodel/SequenceI.html" title="interface in apollo.datamodel">SequenceI</A> <B>peptide</B></PRE>
<DL>
<DL>
</DL>
</DL>

<!-- ========= CONSTRUCTOR DETAIL ======== -->

<A NAME="constructor_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Constructor Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="Transcript()"><!-- --></A><H3>
Transcript</H3>
<PRE>
public <B>Transcript</B>()</PRE>
<DL>
</DL>
<HR>

<A NAME="Transcript(apollo.datamodel.FeatureSetI)"><!-- --></A><H3>
Transcript</H3>
<PRE>
public <B>Transcript</B>(<A HREF="../../apollo/datamodel/FeatureSetI.html" title="interface in apollo.datamodel">FeatureSetI</A>&nbsp;fs)</PRE>
<DL>
</DL>

<!-- ============ METHOD DETAIL ========== -->

<A NAME="method_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Method Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="addExon(apollo.datamodel.ExonI)"><!-- --></A><H3>
addExon</H3>
<PRE>
public void <B>addExon</B>(<A HREF="../../apollo/datamodel/ExonI.html" title="interface in apollo.datamodel">ExonI</A>&nbsp;exon)</PRE>
<DL>
<DD>If exon overlaps with other exons then it merges them, otherwise
      adds exon to transcript
<P>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="sortExons()"><!-- --></A><H3>
sortExons</H3>
<PRE>
public void <B>sortExons</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="addFeature(apollo.datamodel.SeqFeatureI)"><!-- --></A><H3>
addFeature</H3>
<PRE>
public void <B>addFeature</B>(<A HREF="../../apollo/datamodel/SeqFeatureI.html" title="interface in apollo.datamodel">SeqFeatureI</A>&nbsp;feature)</PRE>
<DL>
<DD><B>Description copied from class: <CODE><A HREF="../../apollo/datamodel/FeatureSet.html#addFeature(apollo.datamodel.SeqFeatureI)">FeatureSet</A></CODE></B></DD>
<DD>Add feature to end of features list, recalc low and high
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../apollo/datamodel/SeqFeatureI.html#addFeature(apollo.datamodel.SeqFeatureI)">addFeature</A></CODE> in interface <CODE><A HREF="../../apollo/datamodel/SeqFeatureI.html" title="interface in apollo.datamodel">SeqFeatureI</A></CODE><DT><B>Overrides:</B><DD><CODE><A HREF="../../apollo/datamodel/FeatureSet.html#addFeature(apollo.datamodel.SeqFeatureI)">addFeature</A></CODE> in class <CODE><A HREF="../../apollo/datamodel/FeatureSet.html" title="class in apollo.datamodel">FeatureSet</A></CODE></DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="deleteExon(apollo.datamodel.ExonI)"><!-- --></A><H3>
deleteExon</H3>
<PRE>
public void <B>deleteExon</B>(<A HREF="../../apollo/datamodel/ExonI.html" title="interface in apollo.datamodel">ExonI</A>&nbsp;exon)</PRE>
<DL>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="getExons()"><!-- --></A><H3>
getExons</H3>
<PRE>
public java.util.Vector <B>getExons</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="getExonAt(int)"><!-- --></A><H3>
getExonAt</H3>
<PRE>
public <A HREF="../../apollo/datamodel/ExonI.html" title="interface in apollo.datamodel">ExonI</A> <B>getExonAt</B>(int&nbsp;i)</PRE>
<DL>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="get_cDNA()"><!-- --></A><H3>
get_cDNA</H3>
<PRE>
public java.lang.String <B>get_cDNA</B>()</PRE>
<DL>
<DD>Calls FeatureSet.get_cDNA which appends exon sequence
      (with getSplicedTranscript) where the actual sequence is retrieved 
      from the ref sequence. In the case of overhanging transcripts this
      wont work because there is only ref sequence for part of the transcript.
      So in this case it does not retrieve ref sequence but just returns the
      cdna sequence originally assigend. This is ok because overhanging 
      transcripts are uneditable in apollo, so theres no need to retrieve
      from the ref seq.
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../apollo/datamodel/SeqFeatureI.html#get_cDNA()">get_cDNA</A></CODE> in interface <CODE><A HREF="../../apollo/datamodel/SeqFeatureI.html" title="interface in apollo.datamodel">SeqFeatureI</A></CODE><DT><B>Overrides:</B><DD><CODE><A HREF="../../apollo/datamodel/FeatureSet.html#get_cDNA()">get_cDNA</A></CODE> in class <CODE><A HREF="../../apollo/datamodel/FeatureSet.html" title="class in apollo.datamodel">FeatureSet</A></CODE></DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="haveWholeSequence()"><!-- --></A><H3>
haveWholeSequence</H3>
<PRE>
public boolean <B>haveWholeSequence</B>()</PRE>
<DL>
<DD>Returns true if transcript is hanging off the bounds of the entry
      Does this belong in FeatureSet? Or even SeqFeature?
      Delete this - replaced by isContainedByRefSeq
<P>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="get_cDNASequence()"><!-- --></A><H3>
get_cDNASequence</H3>
<PRE>
public <A HREF="../../apollo/datamodel/SequenceI.html" title="interface in apollo.datamodel">SequenceI</A> <B>get_cDNASequence</B>()</PRE>
<DL>
<DD><B>Description copied from class: <CODE><A HREF="../../apollo/datamodel/AnnotatedFeature.html#get_cDNASequence()">AnnotatedFeature</A></CODE></B></DD>
<DD>by default return null - overridden by Transcript subclass
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../apollo/datamodel/AnnotatedFeatureI.html#get_cDNASequence()">get_cDNASequence</A></CODE> in interface <CODE><A HREF="../../apollo/datamodel/AnnotatedFeatureI.html" title="interface in apollo.datamodel">AnnotatedFeatureI</A></CODE><DT><B>Overrides:</B><DD><CODE><A HREF="../../apollo/datamodel/AnnotatedFeature.html#get_cDNASequence()">get_cDNASequence</A></CODE> in class <CODE><A HREF="../../apollo/datamodel/AnnotatedFeature.html" title="class in apollo.datamodel">AnnotatedFeature</A></CODE></DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="set_cDNASequence(apollo.datamodel.SequenceI)"><!-- --></A><H3>
set_cDNASequence</H3>
<PRE>
public void <B>set_cDNASequence</B>(<A HREF="../../apollo/datamodel/SequenceI.html" title="interface in apollo.datamodel">SequenceI</A>&nbsp;cDNA)</PRE>
<DL>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="translate()"><!-- --></A><H3>
translate</H3>
<PRE>
public java.lang.String <B>translate</B>()</PRE>
<DL>
<DD>Creates SequenceI peptide if doesnt exist already - if translation is null
 or "" peptide will not be created - should it anyways?
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../apollo/datamodel/SeqFeatureI.html#translate()">translate</A></CODE> in interface <CODE><A HREF="../../apollo/datamodel/SeqFeatureI.html" title="interface in apollo.datamodel">SeqFeatureI</A></CODE><DT><B>Overrides:</B><DD><CODE><A HREF="../../apollo/datamodel/FeatureSet.html#translate()">translate</A></CODE> in class <CODE><A HREF="../../apollo/datamodel/FeatureSet.html" title="class in apollo.datamodel">FeatureSet</A></CODE></DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="getPeptideSequence()"><!-- --></A><H3>
getPeptideSequence</H3>
<PRE>
public <A HREF="../../apollo/datamodel/SequenceI.html" title="interface in apollo.datamodel">SequenceI</A> <B>getPeptideSequence</B>()</PRE>
<DL>
<DD>creates peptide sequence if dont have one && prot coding gene - also if pep
 is invalid retranslates
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../apollo/datamodel/SeqFeatureI.html#getPeptideSequence()">getPeptideSequence</A></CODE> in interface <CODE><A HREF="../../apollo/datamodel/SeqFeatureI.html" title="interface in apollo.datamodel">SeqFeatureI</A></CODE><DT><B>Overrides:</B><DD><CODE><A HREF="../../apollo/datamodel/SeqFeature.html#getPeptideSequence()">getPeptideSequence</A></CODE> in class <CODE><A HREF="../../apollo/datamodel/SeqFeature.html" title="class in apollo.datamodel">SeqFeature</A></CODE></DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="setPeptideSequence(apollo.datamodel.SequenceI)"><!-- --></A><H3>
setPeptideSequence</H3>
<PRE>
public void <B>setPeptideSequence</B>(<A HREF="../../apollo/datamodel/SequenceI.html" title="interface in apollo.datamodel">SequenceI</A>&nbsp;pep)</PRE>
<DL>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="setPeptideValidity(boolean)"><!-- --></A><H3>
setPeptideValidity</H3>
<PRE>
public void <B>setPeptideValidity</B>(boolean&nbsp;validity)</PRE>
<DL>
<DD>I think the idea here is if validity is false its saying the peptide is no
 longer valid and a new translation needs to be calculated - this is now
   part of TranslationI - move this method to FeatureSet? (cdna? peptide?)
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../apollo/datamodel/TranslationI.html#setPeptideValidity(boolean)">setPeptideValidity</A></CODE> in interface <CODE><A HREF="../../apollo/datamodel/TranslationI.html" title="interface in apollo.datamodel">TranslationI</A></CODE><DT><B>Overrides:</B><DD><CODE><A HREF="../../apollo/datamodel/FeatureSet.html#setPeptideValidity(boolean)">setPeptideValidity</A></CODE> in class <CODE><A HREF="../../apollo/datamodel/FeatureSet.html" title="class in apollo.datamodel">FeatureSet</A></CODE></DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="getPeptideValidity()"><!-- --></A><H3>
getPeptideValidity</H3>
<PRE>
public boolean <B>getPeptideValidity</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="toString()"><!-- --></A><H3>
toString</H3>
<PRE>
public java.lang.String <B>toString</B>()</PRE>
<DL>
<DD><B>Description copied from class: <CODE><A HREF="../../apollo/datamodel/FeatureSet.html#toString()">FeatureSet</A></CODE></B></DD>
<DD>For debugging
<P>
<DD><DL>
<DT><B>Overrides:</B><DD><CODE><A HREF="../../apollo/datamodel/AnnotatedFeature.html#toString()">toString</A></CODE> in class <CODE><A HREF="../../apollo/datamodel/AnnotatedFeature.html" title="class in apollo.datamodel">AnnotatedFeature</A></CODE></DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="sortTransSpliced()"><!-- --></A><H3>
sortTransSpliced</H3>
<PRE>
public void <B>sortTransSpliced</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="hasPeptideSequence()"><!-- --></A><H3>
hasPeptideSequence</H3>
<PRE>
public boolean <B>hasPeptideSequence</B>()</PRE>
<DL>
<DD>If we have a translation start(nonzero) we have a peptide seq 
      do i have this right? translate may still return null if we have start
      but no transcript sequence(if theres no sequence for the entry). 
      This calls getPeptideSequence() which will actually create a peptide
      sequence if it can & there isnt one yet 
      this aint right - can still have peptide with no trans start - just missing
      5 prime thats all
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../apollo/datamodel/SeqFeatureI.html#hasPeptideSequence()">hasPeptideSequence</A></CODE> in interface <CODE><A HREF="../../apollo/datamodel/SeqFeatureI.html" title="interface in apollo.datamodel">SeqFeatureI</A></CODE><DT><B>Overrides:</B><DD><CODE><A HREF="../../apollo/datamodel/SeqFeature.html#hasPeptideSequence()">hasPeptideSequence</A></CODE> in class <CODE><A HREF="../../apollo/datamodel/SeqFeature.html" title="class in apollo.datamodel">SeqFeature</A></CODE></DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="getTopLevelType()"><!-- --></A><H3>
getTopLevelType</H3>
<PRE>
public java.lang.String <B>getTopLevelType</B>()</PRE>
<DL>
<DD>Transcript has the same biotype as its gene
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../apollo/datamodel/SeqFeatureI.html#getTopLevelType()">getTopLevelType</A></CODE> in interface <CODE><A HREF="../../apollo/datamodel/SeqFeatureI.html" title="interface in apollo.datamodel">SeqFeatureI</A></CODE><DT><B>Overrides:</B><DD><CODE><A HREF="../../apollo/datamodel/SeqFeature.html#getTopLevelType()">getTopLevelType</A></CODE> in class <CODE><A HREF="../../apollo/datamodel/SeqFeature.html" title="class in apollo.datamodel">SeqFeature</A></CODE></DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="getGene()"><!-- --></A><H3>
getGene</H3>
<PRE>
public <A HREF="../../apollo/datamodel/AnnotatedFeatureI.html" title="interface in apollo.datamodel">AnnotatedFeatureI</A> <B>getGene</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="isProteinCodingGene()"><!-- --></A><H3>
isProteinCodingGene</H3>
<PRE>
public boolean <B>isProteinCodingGene</B>()</PRE>
<DL>
<DD><B>Description copied from class: <CODE><A HREF="../../apollo/datamodel/AnnotatedFeature.html#isProteinCodingGene()">AnnotatedFeature</A></CODE></B></DD>
<DD>This is the isProteinCodingGene method that really matters--the ones in other classes
      aren't really used.
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../apollo/datamodel/FeatureSetI.html#isProteinCodingGene()">isProteinCodingGene</A></CODE> in interface <CODE><A HREF="../../apollo/datamodel/FeatureSetI.html" title="interface in apollo.datamodel">FeatureSetI</A></CODE><DT><B>Specified by:</B><DD><CODE><A HREF="../../apollo/datamodel/SeqFeatureI.html#isProteinCodingGene()">isProteinCodingGene</A></CODE> in interface <CODE><A HREF="../../apollo/datamodel/SeqFeatureI.html" title="interface in apollo.datamodel">SeqFeatureI</A></CODE><DT><B>Overrides:</B><DD><CODE><A HREF="../../apollo/datamodel/AnnotatedFeature.html#isProteinCodingGene()">isProteinCodingGene</A></CODE> in class <CODE><A HREF="../../apollo/datamodel/AnnotatedFeature.html" title="class in apollo.datamodel">AnnotatedFeature</A></CODE></DL>
</DD>
<DD><DL>
<DT><B>See Also:</B><DD><A HREF="../../apollo/datamodel/FeatureSet.html#setProteinCodingGene(boolean)"><CODE>FeatureSet.setProteinCodingGene(boolean)</CODE></A></DL>
</DD>
</DL>
<HR>

<A NAME="getNonConsensusAcceptorNum()"><!-- --></A><H3>
getNonConsensusAcceptorNum</H3>
<PRE>
public int <B>getNonConsensusAcceptorNum</B>()</PRE>
<DL>
<DD>This returns the exon index of the non-consensus acceptor (if
      any), or -1 if there aren't any non-consensus acceptors.  I
      think it should be called getNonConsensusAcceptorIndex instead--
      getNonConsensusAcceptorNum makes it sounds like it's returning
      the total number (0 or more) of non-consensus acceptors.
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../apollo/datamodel/AnnotatedFeatureI.html#getNonConsensusAcceptorNum()">getNonConsensusAcceptorNum</A></CODE> in interface <CODE><A HREF="../../apollo/datamodel/AnnotatedFeatureI.html" title="interface in apollo.datamodel">AnnotatedFeatureI</A></CODE><DT><B>Overrides:</B><DD><CODE><A HREF="../../apollo/datamodel/AnnotatedFeature.html#getNonConsensusAcceptorNum()">getNonConsensusAcceptorNum</A></CODE> in class <CODE><A HREF="../../apollo/datamodel/AnnotatedFeature.html" title="class in apollo.datamodel">AnnotatedFeature</A></CODE></DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="getNonConsensusDonorNum()"><!-- --></A><H3>
getNonConsensusDonorNum</H3>
<PRE>
public int <B>getNonConsensusDonorNum</B>()</PRE>
<DL>
<DD><B>Description copied from class: <CODE><A HREF="../../apollo/datamodel/AnnotatedFeature.html#getNonConsensusDonorNum()">AnnotatedFeature</A></CODE></B></DD>
<DD>Returns transcript exon number with non consensus donor. Returns -1 by
      default - Transcript overrides.
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../apollo/datamodel/AnnotatedFeatureI.html#getNonConsensusDonorNum()">getNonConsensusDonorNum</A></CODE> in interface <CODE><A HREF="../../apollo/datamodel/AnnotatedFeatureI.html" title="interface in apollo.datamodel">AnnotatedFeatureI</A></CODE><DT><B>Overrides:</B><DD><CODE><A HREF="../../apollo/datamodel/AnnotatedFeature.html#getNonConsensusDonorNum()">getNonConsensusDonorNum</A></CODE> in class <CODE><A HREF="../../apollo/datamodel/AnnotatedFeature.html" title="class in apollo.datamodel">AnnotatedFeature</A></CODE></DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="nonConsensusSplicingOkay()"><!-- --></A><H3>
nonConsensusSplicingOkay</H3>
<PRE>
public boolean <B>nonConsensusSplicingOkay</B>()</PRE>
<DL>
<DD><B>Description copied from class: <CODE><A HREF="../../apollo/datamodel/AnnotatedFeature.html#nonConsensusSplicingOkay()">AnnotatedFeature</A></CODE></B></DD>
<DD>Return true if non consensus splicing has been deemed ok. This is set
      by curator with method that takes boolean. Overridden by Transcript.
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../apollo/datamodel/AnnotatedFeatureI.html#nonConsensusSplicingOkay()">nonConsensusSplicingOkay</A></CODE> in interface <CODE><A HREF="../../apollo/datamodel/AnnotatedFeatureI.html" title="interface in apollo.datamodel">AnnotatedFeatureI</A></CODE><DT><B>Overrides:</B><DD><CODE><A HREF="../../apollo/datamodel/AnnotatedFeature.html#nonConsensusSplicingOkay()">nonConsensusSplicingOkay</A></CODE> in class <CODE><A HREF="../../apollo/datamodel/AnnotatedFeature.html" title="class in apollo.datamodel">AnnotatedFeature</A></CODE></DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="nonConsensusSplicingOkay(boolean)"><!-- --></A><H3>
nonConsensusSplicingOkay</H3>
<PRE>
public void <B>nonConsensusSplicingOkay</B>(boolean&nbsp;okay)</PRE>
<DL>
<DD><B>Description copied from class: <CODE><A HREF="../../apollo/datamodel/AnnotatedFeature.html#nonConsensusSplicingOkay(boolean)">AnnotatedFeature</A></CODE></B></DD>
<DD>Set whether its ok to have non-consensus splice site - no-op overridden
      by Transcript
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../apollo/datamodel/AnnotatedFeatureI.html#nonConsensusSplicingOkay(boolean)">nonConsensusSplicingOkay</A></CODE> in interface <CODE><A HREF="../../apollo/datamodel/AnnotatedFeatureI.html" title="interface in apollo.datamodel">AnnotatedFeatureI</A></CODE><DT><B>Overrides:</B><DD><CODE><A HREF="../../apollo/datamodel/AnnotatedFeature.html#nonConsensusSplicingOkay(boolean)">nonConsensusSplicingOkay</A></CODE> in class <CODE><A HREF="../../apollo/datamodel/AnnotatedFeature.html" title="class in apollo.datamodel">AnnotatedFeature</A></CODE></DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="isTranscript()"><!-- --></A><H3>
isTranscript</H3>
<PRE>
public boolean <B>isTranscript</B>()</PRE>
<DL>
<DD>Returns true becuase we have a transcript - its ourself. Rename isTranscript? 
      does transcript always have translation start and stop?
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../apollo/datamodel/SeqFeatureI.html#isTranscript()">isTranscript</A></CODE> in interface <CODE><A HREF="../../apollo/datamodel/SeqFeatureI.html" title="interface in apollo.datamodel">SeqFeatureI</A></CODE><DT><B>Overrides:</B><DD><CODE><A HREF="../../apollo/datamodel/SeqFeature.html#isTranscript()">isTranscript</A></CODE> in class <CODE><A HREF="../../apollo/datamodel/SeqFeature.html" title="class in apollo.datamodel">SeqFeature</A></CODE></DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="getProteinFeat()"><!-- --></A><H3>
getProteinFeat</H3>
<PRE>
public <A HREF="../../apollo/datamodel/Protein.html" title="class in apollo.datamodel">Protein</A> <B>getProteinFeat</B>()</PRE>
<DL>
<DD>first cut - eventually grow into full fledged protein feat
      that has a pep seq - which will replace Transcripts peptide Seq
      probably will require a Protein class. returns null if !isProteinCodingGene()
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../apollo/datamodel/SeqFeatureI.html#getProteinFeat()">getProteinFeat</A></CODE> in interface <CODE><A HREF="../../apollo/datamodel/SeqFeatureI.html" title="interface in apollo.datamodel">SeqFeatureI</A></CODE><DT><B>Overrides:</B><DD><CODE><A HREF="../../apollo/datamodel/SeqFeature.html#getProteinFeat()">getProteinFeat</A></CODE> in class <CODE><A HREF="../../apollo/datamodel/SeqFeature.html" title="class in apollo.datamodel">SeqFeature</A></CODE></DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="hasProteinFeat()"><!-- --></A><H3>
hasProteinFeat</H3>
<PRE>
public boolean <B>hasProteinFeat</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
</DL>
<!-- ========= END OF CLASS DATA ========= -->
<HR>


<!-- ======= START OF BOTTOM NAVBAR ====== -->
<A NAME="navbar_bottom"><!-- --></A>
<A HREF="#skip-navbar_bottom" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_bottom_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Class</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;<A HREF="../../apollo/datamodel/SyntenyRegion.html" title="class in apollo.datamodel"><B>PREV CLASS</B></A>&nbsp;
&nbsp;<A HREF="../../apollo/datamodel/TranslationI.html" title="interface in apollo.datamodel"><B>NEXT CLASS</B></A></FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../index.html?apollo/datamodel/Transcript.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="Transcript.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
<TR>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
  SUMMARY:&nbsp;NESTED&nbsp;|&nbsp;<A HREF="#field_summary">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_summary">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_summary">METHOD</A></FONT></TD>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
DETAIL:&nbsp;<A HREF="#field_detail">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_detail">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_detail">METHOD</A></FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_bottom"></A>
<!-- ======== END OF BOTTOM NAVBAR ======= -->

<HR>

</BODY>
</HTML>
