<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!--NewPage-->
<HTML>
<HEAD>
<!-- Generated by javadoc (build 1.5.0_18) on Tue Jul 07 10:40:04 PDT 2009 -->
<TITLE>
DataFetcher
</TITLE>

<META NAME="keywords" CONTENT="apollo.dataadapter.ensj.DataFetcher class">

<LINK REL ="stylesheet" TYPE="text/css" HREF="../../../stylesheet.css" TITLE="Style">

<SCRIPT type="text/javascript">
function windowTitle()
{
    parent.document.title="DataFetcher";
}
</SCRIPT>
<NOSCRIPT>
</NOSCRIPT>

</HEAD>

<BODY BGCOLOR="white" onload="windowTitle();">


<!-- ========= START OF TOP NAVBAR ======= -->
<A NAME="navbar_top"><!-- --></A>
<A HREF="#skip-navbar_top" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_top_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Class</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;<A HREF="../../../apollo/dataadapter/ensj/AlternateEnsJSequence.html" title="class in apollo.dataadapter.ensj"><B>PREV CLASS</B></A>&nbsp;
&nbsp;<A HREF="../../../apollo/dataadapter/ensj/DataModelConversionUtil.html" title="class in apollo.dataadapter.ensj"><B>NEXT CLASS</B></A></FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../../index.html?apollo/dataadapter/ensj/DataFetcher.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="DataFetcher.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
<TR>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
  SUMMARY:&nbsp;NESTED&nbsp;|&nbsp;<A HREF="#field_summary">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_summary">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_summary">METHOD</A></FONT></TD>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
DETAIL:&nbsp;<A HREF="#field_detail">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_detail">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_detail">METHOD</A></FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_top"></A>
<!-- ========= END OF TOP NAVBAR ========= -->

<HR>
<!-- ======== START OF CLASS DATA ======== -->
<H2>
<FONT SIZE="-1">
apollo.dataadapter.ensj</FONT>
<BR>
Class DataFetcher</H2>
<PRE>
java.lang.Object
  <IMG SRC="../../../resources/inherit.gif" ALT="extended by "><B>apollo.dataadapter.ensj.DataFetcher</B>
</PRE>
<HR>
<DL>
<DT><PRE>public class <B>DataFetcher</B><DT>extends java.lang.Object</DL>
</PRE>

<P>
I am just a separate helper class to stop the EnsJAdapter from becoming too huge.
 I keep all the code that drives ensj to fetch stuff.
<P>

<P>
<HR>

<P>
<!-- =========== FIELD SUMMARY =========== -->

<A NAME="field_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Field Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected static&nbsp;org.apache.log4j.Logger</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../apollo/dataadapter/ensj/DataFetcher.html#logger">logger</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<!-- ======== CONSTRUCTOR SUMMARY ======== -->

<A NAME="constructor_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Constructor Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../apollo/dataadapter/ensj/DataFetcher.html#DataFetcher(apollo.dataadapter.ensj.EnsJAdapter)">DataFetcher</A></B>(<A HREF="../../../apollo/dataadapter/ensj/EnsJAdapter.html" title="class in apollo.dataadapter.ensj">EnsJAdapter</A>&nbsp;adapter)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<!-- ========== METHOD SUMMARY =========== -->

<A NAME="method_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Method Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../apollo/dataadapter/ensj/DataFetcher.html#addContigFeatures(apollo.datamodel.CurationSet, java.util.List, org.ensembl.datamodel.Location)">addContigFeatures</A></B>(<A HREF="../../../apollo/datamodel/CurationSet.html" title="class in apollo.datamodel">CurationSet</A>&nbsp;curationSet,
                  java.util.List&nbsp;flatFeatures,
                  org.ensembl.datamodel.Location&nbsp;location)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../apollo/dataadapter/ensj/DataFetcher.html#addDitagFeatures(apollo.datamodel.CurationSet, java.util.List, org.ensembl.datamodel.Location, java.util.List)">addDitagFeatures</A></B>(<A HREF="../../../apollo/datamodel/CurationSet.html" title="class in apollo.datamodel">CurationSet</A>&nbsp;curationSet,
                 java.util.List&nbsp;flatFeatures,
                 org.ensembl.datamodel.Location&nbsp;location,
                 java.util.List&nbsp;includedTypes)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../apollo/dataadapter/ensj/DataFetcher.html#addDnaAlignFeatures(apollo.datamodel.CurationSet, java.util.List, org.ensembl.datamodel.Location, java.util.List)">addDnaAlignFeatures</A></B>(<A HREF="../../../apollo/datamodel/CurationSet.html" title="class in apollo.datamodel">CurationSet</A>&nbsp;curationSet,
                    java.util.List&nbsp;flatFeatures,
                    org.ensembl.datamodel.Location&nbsp;location,
                    java.util.List&nbsp;includedTypes)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.util.List</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../apollo/dataadapter/ensj/DataFetcher.html#addGenes(apollo.datamodel.StrandedFeatureSet, org.ensembl.datamodel.Location, java.util.List, boolean, java.lang.String)">addGenes</A></B>(<A HREF="../../../apollo/datamodel/StrandedFeatureSet.html" title="class in apollo.datamodel">StrandedFeatureSet</A>&nbsp;results,
         org.ensembl.datamodel.Location&nbsp;location,
         java.util.List&nbsp;includedTypes,
         boolean&nbsp;aggressiveNaming,
         java.lang.String&nbsp;typePrefix)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Retrieves ensJ genes/transcripts/exons, converts them into seq features and adds them
 to the curation set.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../apollo/dataadapter/ensj/DataFetcher.html#addPredictionTranscripts(apollo.datamodel.StrandedFeatureSet, org.ensembl.datamodel.Location, java.util.List)">addPredictionTranscripts</A></B>(<A HREF="../../../apollo/datamodel/StrandedFeatureSet.html" title="class in apollo.datamodel">StrandedFeatureSet</A>&nbsp;results,
                         org.ensembl.datamodel.Location&nbsp;location,
                         java.util.List&nbsp;includedTypes)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../apollo/dataadapter/ensj/DataFetcher.html#addProteinAlignFeatures(apollo.datamodel.CurationSet, java.util.List, org.ensembl.datamodel.Location, java.util.List)">addProteinAlignFeatures</A></B>(<A HREF="../../../apollo/datamodel/CurationSet.html" title="class in apollo.datamodel">CurationSet</A>&nbsp;curationSet,
                        java.util.List&nbsp;flatFeatures,
                        org.ensembl.datamodel.Location&nbsp;location,
                        java.util.List&nbsp;includedTypes)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../apollo/dataadapter/ensj/DataFetcher.html#addProteinAnnotations(apollo.datamodel.CurationSet, java.util.List, java.util.List)">addProteinAnnotations</A></B>(<A HREF="../../../apollo/datamodel/CurationSet.html" title="class in apollo.datamodel">CurationSet</A>&nbsp;curationSet,
                      java.util.List&nbsp;flatFeatures,
                      java.util.List&nbsp;retrievedGenes)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../apollo/dataadapter/ensj/DataFetcher.html#addRepeatFeatures(apollo.datamodel.CurationSet, java.util.List, org.ensembl.datamodel.Location)">addRepeatFeatures</A></B>(<A HREF="../../../apollo/datamodel/CurationSet.html" title="class in apollo.datamodel">CurationSet</A>&nbsp;curationSet,
                  java.util.List&nbsp;flatFeatures,
                  org.ensembl.datamodel.Location&nbsp;location)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Add all repeat features into the input set.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../apollo/dataadapter/ensj/DataFetcher.html#addSimpleFeatures(apollo.datamodel.CurationSet, java.util.List, org.ensembl.datamodel.Location, java.util.List)">addSimpleFeatures</A></B>(<A HREF="../../../apollo/datamodel/CurationSet.html" title="class in apollo.datamodel">CurationSet</A>&nbsp;curationSet,
                  java.util.List&nbsp;flatFeatures,
                  org.ensembl.datamodel.Location&nbsp;location,
                  java.util.List&nbsp;includedTypes)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;<A HREF="../../../apollo/datamodel/StrandedFeatureSetI.html" title="interface in apollo.datamodel">StrandedFeatureSetI</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../apollo/dataadapter/ensj/DataFetcher.html#getAnnotations(apollo.datamodel.CurationSet, java.util.List, boolean)">getAnnotations</A></B>(<A HREF="../../../apollo/datamodel/CurationSet.html" title="class in apollo.datamodel">CurationSet</A>&nbsp;curationSet,
               java.util.List&nbsp;genes,
               boolean&nbsp;addTranscriptSupport)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;<A NAME="methods_inherited_from_class_java.lang.Object"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left"><B>Methods inherited from class java.lang.Object</B></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE>clone, equals, finalize, getClass, hashCode, notify, notifyAll, toString, wait, wait, wait</CODE></TD>
</TR>
</TABLE>
&nbsp;
<P>

<!-- ============ FIELD DETAIL =========== -->

<A NAME="field_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Field Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="logger"><!-- --></A><H3>
logger</H3>
<PRE>
protected static final org.apache.log4j.Logger <B>logger</B></PRE>
<DL>
<DL>
</DL>
</DL>

<!-- ========= CONSTRUCTOR DETAIL ======== -->

<A NAME="constructor_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Constructor Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="DataFetcher(apollo.dataadapter.ensj.EnsJAdapter)"><!-- --></A><H3>
DataFetcher</H3>
<PRE>
public <B>DataFetcher</B>(<A HREF="../../../apollo/dataadapter/ensj/EnsJAdapter.html" title="class in apollo.dataadapter.ensj">EnsJAdapter</A>&nbsp;adapter)</PRE>
<DL>
</DL>

<!-- ============ METHOD DETAIL ========== -->

<A NAME="method_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Method Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="addGenes(apollo.datamodel.StrandedFeatureSet, org.ensembl.datamodel.Location, java.util.List, boolean, java.lang.String)"><!-- --></A><H3>
addGenes</H3>
<PRE>
public java.util.List <B>addGenes</B>(<A HREF="../../../apollo/datamodel/StrandedFeatureSet.html" title="class in apollo.datamodel">StrandedFeatureSet</A>&nbsp;results,
                               org.ensembl.datamodel.Location&nbsp;location,
                               java.util.List&nbsp;includedTypes,
                               boolean&nbsp;aggressiveNaming,
                               java.lang.String&nbsp;typePrefix)
                        throws org.ensembl.driver.AdaptorException</PRE>
<DL>
<DD>Retrieves ensJ genes/transcripts/exons, converts them into seq features and adds them
 to the curation set. Passes out the list of ensj-genes retreived: this so that the simple-feature
 adapter can, if it wants, use the translations to retrieve protein annotations.
<P>
<DD><DL>

<DT><B>Throws:</B>
<DD><CODE>org.ensembl.driver.AdaptorException</CODE></DL>
</DD>
</DL>
<HR>

<A NAME="addRepeatFeatures(apollo.datamodel.CurationSet, java.util.List, org.ensembl.datamodel.Location)"><!-- --></A><H3>
addRepeatFeatures</H3>
<PRE>
public int <B>addRepeatFeatures</B>(<A HREF="../../../apollo/datamodel/CurationSet.html" title="class in apollo.datamodel">CurationSet</A>&nbsp;curationSet,
                             java.util.List&nbsp;flatFeatures,
                             org.ensembl.datamodel.Location&nbsp;location)
                      throws org.ensembl.driver.AdaptorException</PRE>
<DL>
<DD>Add all repeat features into the input set.
<P>
<DD><DL>

<DT><B>Throws:</B>
<DD><CODE>org.ensembl.driver.AdaptorException</CODE></DL>
</DD>
</DL>
<HR>

<A NAME="addProteinAlignFeatures(apollo.datamodel.CurationSet, java.util.List, org.ensembl.datamodel.Location, java.util.List)"><!-- --></A><H3>
addProteinAlignFeatures</H3>
<PRE>
public int <B>addProteinAlignFeatures</B>(<A HREF="../../../apollo/datamodel/CurationSet.html" title="class in apollo.datamodel">CurationSet</A>&nbsp;curationSet,
                                   java.util.List&nbsp;flatFeatures,
                                   org.ensembl.datamodel.Location&nbsp;location,
                                   java.util.List&nbsp;includedTypes)
                            throws org.ensembl.driver.AdaptorException</PRE>
<DL>
<DD><DL>

<DT><B>Throws:</B>
<DD><CODE>org.ensembl.driver.AdaptorException</CODE></DL>
</DD>
</DL>
<HR>

<A NAME="addDnaAlignFeatures(apollo.datamodel.CurationSet, java.util.List, org.ensembl.datamodel.Location, java.util.List)"><!-- --></A><H3>
addDnaAlignFeatures</H3>
<PRE>
public int <B>addDnaAlignFeatures</B>(<A HREF="../../../apollo/datamodel/CurationSet.html" title="class in apollo.datamodel">CurationSet</A>&nbsp;curationSet,
                               java.util.List&nbsp;flatFeatures,
                               org.ensembl.datamodel.Location&nbsp;location,
                               java.util.List&nbsp;includedTypes)
                        throws org.ensembl.driver.AdaptorException</PRE>
<DL>
<DD><DL>

<DT><B>Throws:</B>
<DD><CODE>org.ensembl.driver.AdaptorException</CODE></DL>
</DD>
</DL>
<HR>

<A NAME="addDitagFeatures(apollo.datamodel.CurationSet, java.util.List, org.ensembl.datamodel.Location, java.util.List)"><!-- --></A><H3>
addDitagFeatures</H3>
<PRE>
public int <B>addDitagFeatures</B>(<A HREF="../../../apollo/datamodel/CurationSet.html" title="class in apollo.datamodel">CurationSet</A>&nbsp;curationSet,
                            java.util.List&nbsp;flatFeatures,
                            org.ensembl.datamodel.Location&nbsp;location,
                            java.util.List&nbsp;includedTypes)
                     throws org.ensembl.driver.AdaptorException</PRE>
<DL>
<DD><DL>

<DT><B>Throws:</B>
<DD><CODE>org.ensembl.driver.AdaptorException</CODE></DL>
</DD>
</DL>
<HR>

<A NAME="addPredictionTranscripts(apollo.datamodel.StrandedFeatureSet, org.ensembl.datamodel.Location, java.util.List)"><!-- --></A><H3>
addPredictionTranscripts</H3>
<PRE>
public int <B>addPredictionTranscripts</B>(<A HREF="../../../apollo/datamodel/StrandedFeatureSet.html" title="class in apollo.datamodel">StrandedFeatureSet</A>&nbsp;results,
                                    org.ensembl.datamodel.Location&nbsp;location,
                                    java.util.List&nbsp;includedTypes)
                             throws org.ensembl.driver.AdaptorException</PRE>
<DL>
<DD><DL>

<DT><B>Throws:</B>
<DD><CODE>org.ensembl.driver.AdaptorException</CODE></DL>
</DD>
</DL>
<HR>

<A NAME="addSimpleFeatures(apollo.datamodel.CurationSet, java.util.List, org.ensembl.datamodel.Location, java.util.List)"><!-- --></A><H3>
addSimpleFeatures</H3>
<PRE>
public int <B>addSimpleFeatures</B>(<A HREF="../../../apollo/datamodel/CurationSet.html" title="class in apollo.datamodel">CurationSet</A>&nbsp;curationSet,
                             java.util.List&nbsp;flatFeatures,
                             org.ensembl.datamodel.Location&nbsp;location,
                             java.util.List&nbsp;includedTypes)
                      throws org.ensembl.driver.AdaptorException</PRE>
<DL>
<DD><DL>

<DT><B>Throws:</B>
<DD><CODE>org.ensembl.driver.AdaptorException</CODE></DL>
</DD>
</DL>
<HR>

<A NAME="addContigFeatures(apollo.datamodel.CurationSet, java.util.List, org.ensembl.datamodel.Location)"><!-- --></A><H3>
addContigFeatures</H3>
<PRE>
public int <B>addContigFeatures</B>(<A HREF="../../../apollo/datamodel/CurationSet.html" title="class in apollo.datamodel">CurationSet</A>&nbsp;curationSet,
                             java.util.List&nbsp;flatFeatures,
                             org.ensembl.datamodel.Location&nbsp;location)
                      throws org.ensembl.driver.AdaptorException</PRE>
<DL>
<DD><DL>

<DT><B>Throws:</B>
<DD><CODE>org.ensembl.driver.AdaptorException</CODE></DL>
</DD>
</DL>
<HR>

<A NAME="addProteinAnnotations(apollo.datamodel.CurationSet, java.util.List, java.util.List)"><!-- --></A><H3>
addProteinAnnotations</H3>
<PRE>
public int <B>addProteinAnnotations</B>(<A HREF="../../../apollo/datamodel/CurationSet.html" title="class in apollo.datamodel">CurationSet</A>&nbsp;curationSet,
                                 java.util.List&nbsp;flatFeatures,
                                 java.util.List&nbsp;retrievedGenes)
                          throws org.ensembl.driver.AdaptorException</PRE>
<DL>
<DD><DL>

<DT><B>Throws:</B>
<DD><CODE>org.ensembl.driver.AdaptorException</CODE></DL>
</DD>
</DL>
<HR>

<A NAME="getAnnotations(apollo.datamodel.CurationSet, java.util.List, boolean)"><!-- --></A><H3>
getAnnotations</H3>
<PRE>
protected <A HREF="../../../apollo/datamodel/StrandedFeatureSetI.html" title="interface in apollo.datamodel">StrandedFeatureSetI</A> <B>getAnnotations</B>(<A HREF="../../../apollo/datamodel/CurationSet.html" title="class in apollo.datamodel">CurationSet</A>&nbsp;curationSet,
                                             java.util.List&nbsp;genes,
                                             boolean&nbsp;addTranscriptSupport)
                                      throws <A HREF="../../../apollo/dataadapter/ApolloAdapterException.html" title="class in apollo.dataadapter">ApolloAdapterException</A>,
                                             org.ensembl.driver.AdaptorException</PRE>
<DL>
<DD><DL>

<DT><B>Throws:</B>
<DD><CODE><A HREF="../../../apollo/dataadapter/ApolloAdapterException.html" title="class in apollo.dataadapter">ApolloAdapterException</A></CODE>
<DD><CODE>org.ensembl.driver.AdaptorException</CODE></DL>
</DD>
</DL>
<!-- ========= END OF CLASS DATA ========= -->
<HR>


<!-- ======= START OF BOTTOM NAVBAR ====== -->
<A NAME="navbar_bottom"><!-- --></A>
<A HREF="#skip-navbar_bottom" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_bottom_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Class</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;<A HREF="../../../apollo/dataadapter/ensj/AlternateEnsJSequence.html" title="class in apollo.dataadapter.ensj"><B>PREV CLASS</B></A>&nbsp;
&nbsp;<A HREF="../../../apollo/dataadapter/ensj/DataModelConversionUtil.html" title="class in apollo.dataadapter.ensj"><B>NEXT CLASS</B></A></FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../../index.html?apollo/dataadapter/ensj/DataFetcher.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="DataFetcher.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
<TR>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
  SUMMARY:&nbsp;NESTED&nbsp;|&nbsp;<A HREF="#field_summary">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_summary">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_summary">METHOD</A></FONT></TD>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
DETAIL:&nbsp;<A HREF="#field_detail">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_detail">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_detail">METHOD</A></FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_bottom"></A>
<!-- ======== END OF BOTTOM NAVBAR ======= -->

<HR>

</BODY>
</HTML>
